docs/apache-airflow/public-airflow-api.rst

```diff
@@ -0,0 +1,141 @@
+The REST API is a set of programming interfaces that allow developers to access certain features of
+Airflow over the HTTP protocol, using a standard set of RESTful (representational state transfer) conventions.
+
+The stable REST API is considered to be part of the public API because it is designed to be used by
+external developers to interact with Airflow in a consistent and predictable way. The REST API provides
+access to many of the same features and functionality that are available through the other public APIs
+of Airflow, such as the ability to create and manage DAGs and tasks, and to monitor their status.
+
+Overall, the stable REST API is an important part of the public API of Apache Airflow, and can
+be a useful tool for developers who want to integrate Airflow with other systems or
+build custom tools and applications on top of Airflow.
+
+Which classes and packages are a public API of Apache Airflow?
+==============================================================
+
+The specific classes and packages that are considered to be part of the public API of Apache
+Airflow can vary depending on the version of Airflow that you are using. In general, however, the
+public API of Airflow consists of a number of different classes and packages that provide access
+to the core features and functionality of the system.
+
+Some examples of the classes and packages that may be considered as the public API of Apache Airflow include:
+
+* The :class:`~airflow.models.dag.DAG`, which provides a way to define and manage DAGs in Airflow.
+* The :class:`~airflow.models.baseoperator.BaseOperator`, which provides a way write custom operators.
+* The :class:`~airflow.hooks.base.BaseHook`, which provides a way write custom hooks.
+* The :class:`~airflow.decorators.base.TaskDecorator`, which provides a way write custom hooks.
+* :class:`~airflow.models.connection.Connection`, :class:`~airflow.models.variable.Variable` and
+  :class:`~airflow.models.xcom.XCom` classes that are used to access and manage environment of execution of
+  the tasks and allow to perform an inter-task communication.
+* :class:`~airflow.executors.base_executor.BaseExecutor` - the Executors are the components of
+  Airflow that are responsible for executing tasks. There are a number of different executor implementations
+  available in Airflow, each with its own unique characteristics and capabilities. You can develop your own
+  executors in the way that suits your needs.
+
+.. note::
+
+    Originally Executors had a number of internal details and coupling with internal Airflow core logic
+    that made it difficult to write your own executors. But as of
+    Airflow 2.6.0 and `AIP-51 <https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-51+Removing+Executor+Coupling+from+Core+Airflow>`_
+    this is no longer the case. The Executors are now fully decoupled and are a part of the public API.
```

Comment:
I think. It's equivalent of "Added in 2.X.0" note.

People seeing this document will extrapolate it to previous versions. And I think they have the right to do it because - despite of claiming it works, we miserably failed in fulfilling our promise there. We knew people were using and developing custom executors and we suggested they can do it, yet we did not make it "truly" possible. And if people will try to do  it for earlier versions they might fail without knowin why.

While I agree the link and note is a bit "over-the-top". I will just leave

```
.. versionadded:: 2.6
```

Because this is the version where it wil be actually possible to get your custom executor "really" works in a decoupled way.